File: internal/service/frontend/renop-html/js/settings/maven-domains.js
/*
 * Copyright (c) 2026 404Setup. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * If it is not possible or desirable to put the notice in a particular file, then You may include the notice in a location (such as a LICENSE file in a relevant directory) where a recipient would be likely to look for such a notice.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as defined by the Mozilla Public License, v. 2.0.
 */

import {el} from '@renop/ui/dom';
import {makeCustomSelect} from '@renop/ui/custom-select';
import {buildInput, createSection} from '../cfg-ui.js';
import {createFieldRow, createIcon} from '../components.js';
import {t} from '../i18n.js';

/** Render the reservation period applied when a publishing domain becomes unsafe. */
export function renderMavenDomainSettings(container, data, changed) {
    const wrap = el('div', {class: 'cfg-layout'});
    const section = createSection(createIcon('network'), t('maven.healthSettings'), t('maven.healthSettingsHint'), {defaultCollapsed: true});
    section.id = 'settings-maven-domains';
    const value = buildInput('number', data.release_value, '2', event => {
        data.release_value = Number(event.target.value);
        changed();
    });
    value.min = '1';
    value.max = '100';
    value.step = '1';
    value.setAttribute('aria-label', t('maven.releasePeriod'));
    const unit = makeCustomSelect(['month', 'year'].map(value => ({value, label: t(`maven.releaseUnit.${value}`)})),
        data.release_unit, unit => {
            data.release_unit = unit;
            changed();
        });
    unit.querySelector('button')?.setAttribute('aria-label', t('mail.unit'));
    section.querySelector('.cfg-fields').append(
        createFieldRow(t('maven.releasePeriod'), '', value), createFieldRow(t('mail.unit'), '', unit));
    wrap.appendChild(section);
    container.appendChild(wrap);
}

File: internal/config/maven_domains.go
/*
 * Copyright (c) 2026 404Setup. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * This Source Code Form is "Incompatible With Secondary Licenses", as defined by the Mozilla Public License, v. 2.0.
 */

package config

import (
	"errors"
	"time"
)

// MavenDomainConfig controls how long a security-locked publishing domain is reserved.
type MavenDomainConfig struct {
	ReleaseValue int    `json:"release_value" yaml:"release_value"`
	ReleaseUnit  string `json:"release_unit" yaml:"release_unit"`
}

// DefaultMavenDomainConfig reserves a locked domain for two calendar years.
func DefaultMavenDomainConfig() MavenDomainConfig {
	return MavenDomainConfig{ReleaseValue: 2, ReleaseUnit: "year"}
}

// Validate rejects disabled, unbounded, or ambiguous release periods.
func (c MavenDomainConfig) Validate() error {
	if c.ReleaseValue < 1 || c.ReleaseValue > 100 || (c.ReleaseUnit != "month" && c.ReleaseUnit != "year") {
		return errors.New("Maven domain release period must be 1 to 100 months or years")
	}
	return nil
}

// ReleaseAt returns the end of the reservation using UTC calendar arithmetic.
func (c MavenDomainConfig) ReleaseAt(lockedAt int64) int64 {
	if c.Validate() != nil || lockedAt <= 0 {
		return 0
	}
	months := c.ReleaseValue
	if c.ReleaseUnit == "year" {
		months *= 12
	}
	return time.UnixMilli(lockedAt).UTC().AddDate(0, months, 0).UnixMilli()
}
